capstone-rs/src/instruction.rs

```diff
@@ -172,6 +225,56 @@ pub struct Insn<'a> {
     pub(crate) _marker: PhantomData<&'a InsnDetail<'a>>,
 }
 
+unsafe impl<'a> Instruction for Insn<'a> {
+    fn insn(&self) -> &cs_insn {
+        &self.insn
+    }
+}
+
+impl<'a> From<&Insn<'_>> for OwnedInsn<'a> {
+    fn from(insn: &Insn<'_>) -> Self {
+        let mut new = unsafe { <*const cs_insn>::read(&insn.insn as _) };
+        new.detail = if new.detail.is_null() {
+            new.detail
+        } else {
+            unsafe {
```

Comment:
Add a comment like this:

```rust
// SAFETY: assumes that `cs_detail` struct transitively only contains owned
// types and no pointers, including the union over the architecture-specific
// types.
```

I looked over the `cs_detail` definition and it seems fine, but we need to make sure that upstream capstone doesn't make "interesting" changes with how `cs_detail` is stored. I've fixed several soundness issues upstream where something weird was being done to shared mutable state without synchronization.

Changing `cs_detail` would be an API/ABI break, so we *should* be fine with upgrades to new minor versions of capstone.